RFC: Runtime Service Module Support (Rejected) (#877)
OpenDevicePartnership/patina: docs/src/rfc/text/0000-runtime-services-support.md

3. Long term goals should remove dependence on EDKII runtime code.

## Unresolved Questions

- Is there a suitable type of dynamic linking that may make other options suitable
  without excessive one-off infrastructure?

## Prior Art (Existing PI C Implementation)

The current C implementation consists of one primary runtime module, usually
RuntimeDxe and a variable amount of additional runtime modules. These modules are
dispatched by DXE, but will be loaded in runtime memory. The primary runtime module
should produce the runtime architecture protocol, defined in the PI spec section
12.8. This protocol allows for the core to fill in critical information for the
runtime module to implement SetVirtualAddressMap. This includes the image list of
runtime modules, events registered for virtualizing memory, and memory map
information that is not currently used. When `SetVirtualAddressMap` is called, the
runtime DXE module will use the information from the protocol it published and the
core filled to re-relocate the images, virtualizing their execution, and call all
events registered to allow drivers/libraries to virtualize any stored pointers.

> ## Alternatives

Reviewer: I may have missed it, but what is the current proposed solution versus listing of alternatives?

Reviewer: Closing so the rejected RFC can be merged

dynamic linking. This would also invent additional infrastructure needed only for
runtime services.

### Runtime Service Removal

The long term goal of runtime services, instead of replacing them, should be to
remove them. Most runtime services are superfluous on modern machines. The only
runtime service that is typically invoked at runtime that cannot be replaced with
other existing mechanisms is variable services. Even so, variable services are nearly
always a simple wrapper around calls to the implementation executing in either SMM
or TrustZone. An alternative to re-implementing runtime services for Patina, would
be to simply not support runtime services at runtime, but instead provide a
specification defined interface to access variable services directly through SMM
and TrustZone.

However, this is not feasible in the short term. Any solution here would require
updates both to the UEFI specifications to support them as well as specification
of the new SMM/TrustZone interfaces, not to mention OS support. While removal of
runtime services is a desirable long term goal it does not satisfy the immediate

> goals of this RFC.

Reviewer 2: Editorial: I've seen prior art in this space (i.e. driver-based variable services with no runtime support from UEFI). Thar be dragons. The main issue is that it breaks a fundamental assumption in much of the code that Variable Services (and other runtime services) exist from the genesis of the OS and are available from the dawn of time - but removing them from runtime implies a blackout period from EBS (where they would presumably cease to function) until such time as whatever OS and driver stack needed to support them arises. 

A strong OS spec on this - i.e "Thou shalt implement TZ or SMM-based runtime services via the following industry standardized interfaces and OS native support and may not provide thine own driver to produce them" would help, I think. 

For completeness, the other place besides var services where UEFI runtime services traditionally have been difficult is clock support for devices that don't have a trivial PC-AT RTC at I/O 0x70/71. This has also proven to be a source of bugs when using e.g. ACPI TimeAndAlarm-driver based solutions that need to co-ordinate/synchronize with UEFI time.

Author: Agreed, we would need some strict requirements defined and public, and the intention is to expose management mode code interface rather than relying on complex OS code. Will resolve this thread barring any concern with the alternative as described.

### Defer Runtime Module Support

Another solution is to defer implementation of any Rust runtime module until DXE
and MM are solved and a more clear interfacing model appears, and continue to use
the EDKII implementation of RuntimeDxe in the interim. The EDKII implementation
would either be replaced by a more robust implementation of a Patina runtime or
removed in the future if runtime services are no longer required in the future.
However, as explained above a more robust solution in this space is filled with
complications and is unlikely to occur in the near future. Additionally, runtime
service removal will be a long process of standardization and OS support. Deferring
until a more complete solution may not be feasible.

## Rust Code Design

Patina should continue to use separate runtime modules which interface using the
UEFI and PI defined interfaces. This includes the system architectural protocol for
resources from the core as well as boot services for allocations and other DXE
phase calls.

> However, as a stepping stone to replacement RuntimeDxe should be implementation of a

Reviewer 3: ```suggestion
However, as a stepping stone to replacing RuntimeDxe, Patina should implement a
```

Rust driver leveraging the original code removed in the above PRs. This driver would
be implemented in the Rust driver infrastructure developed for EDKII and located in
the PatinaPkg. This module would use all existing UEFI/PI interfaces to access boot
services and communicate runtime image and event data. The responsibilities of each
module in this model are detailed below.

### Patina DXE Core

The Patina DXE core will continue to listen for the registration of the runtime
architectural protocol, and register event and image data to the protocol. This
will be updated whenever a new runtime image is loaded or a new `SetVirtualAddressMap`
callback is registered.

### Patina Runtime (new)

> A new module compiled as a Rust driver in an EDKII environment, provides a

Author: For what it's worth, we aren't actually filling any of the memory map fields today as they are not actually used. The only information we are providing is the image and event lists.

Overall, I agree that switching to a rust implementation of RuntimeDxe seems reasonable, and I think its fine to put as the point of record with this RFC, though I'm not concerned with this being accomplished in any immediate timeframe and the cost/reward seems limited.
